pulp_docker/docker_convert.py

```diff
@@ -0,0 +1,256 @@
+        self.compute_layers()
+        manifest = dict(name=name, tag=self.tag, architecture=self.config_layer['architecture'],
+                        schemaVersion=1, fsLayers=self.fs_layers, history=self.history)
+        key = jwk.ECKey().load_key(ecc.P256)
+        key.kid = getKeyId(key)
+        manifData = sign(manifest, key)
+        return manifData
+
+    def compute_layers(self):
+        """
+        Compute layers to be present in the converted image.
+        Empty (throwaway) layers will be created to store image metadata
+        """
+        # Layers in v2s1 are in reverse order from v2s2
+        layers = reversed(self.manifest['layers'])
+        config_layer_history = reversed(self.config_layer['history'])
+        diff_ids = reversed(self.config_layer['rootfs']['diff_ids'])
+        fs_layers = []
+        curr_compressed_dig = next(layers)['digest']
+        curr_uncompressed_dig = next(diff_ids)
+        for curr_hist in config_layer_history:
+            if curr_hist.get("empty_layer"):
+                layer_id = self.EMPTY_LAYER
+                uncompressed_dig = None
+            else:
+                layer_id = curr_compressed_dig
+                uncompressed_dig = curr_uncompressed_dig
+                try:
+                    curr_compressed_dig = next(layers)['digest']
+                    curr_uncompressed_dig = next(diff_ids)
+                except StopIteration:
+                    curr_compressed_dig = self.EMPTY_LAYER
+                    curr_uncompressed_dig = None
+            fs_layers.append((layer_id, uncompressed_dig, curr_hist))
+        self.fs_layers = [dict(blobSum=x[0]) for x in fs_layers]
+        # Compute v1 compatibility
+        parent = None
+        history_entries = self.history = []
+
+        fs_layers_count = len(fs_layers)
+        # Reverse list so we can compute parent/child properly
+        fs_layers.reverse()
+        for i, (compressed_dig, uncompressed_dig, hist) in enumerate(fs_layers):
+            layer_id = self._compute_layer_id(compressed_dig, uncompressed_dig, i)
+            # Last layer?
+            if i == fs_layers_count - 1:
+                # The whole config layer becomes part of the v1compatibility
+                # (minus history and rootfs)
+                config = dict(self.config_layer)
+                config.pop("history", None)
+                config.pop("rootfs", None)
+            else:
+                config = dict(created=hist['created'],
+                              container_config=dict(Cmd=hist['created_by']))
+            if uncompressed_dig is None:
+                config['throwaway'] = True
+            config['id'] = layer_id
+            if parent is not None:
+                config['parent'] = parent
+            parent = layer_id
+            history_entries.append(dict(v1Compatibility=_jsonDumpsCompact(config)))
+        # Reverse again for proper order
+        history_entries.reverse()
```

Comment:
I strongly recommend you to extract at least one more function from the highlighted function. In my opinion, it would improve an overall readability of the code.